### PR TITLE
[Train] Automatically set `NCCL_SOCKET_IFNAME` to use ethernet

### LINKED ISSUE #26663: [AIR][Train] Add default network setting for pytorch multi-host training
### What happened + What you expected to happen

Currently user using AIR to do pytorch distributed training across multiple hosts will run into ganging issue if ens was not set or not set correctly. This hurts dev UX.

In https://github.com/inkawhich/pt-distributed-tutorial/blob/master/pytorch-aws-distributed-tutorial.py it mentioned about using `ipconfig` to find and set right interface with env var `NCCL_SOCKET_IFNAME`

```
ens5: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9001
        inet 172.31.65.108  netmask 255.255.224.0  broadcast 172.31.95.255
        inet6 fe80::865:1bff:fe4c:2e03  prefixlen 64  scopeid 0x20<link>
        ether 0a:65:1b:4c:2e:03  txqueuelen 1000  (Ethernet)
        RX packets 12552852  bytes 18852483513 (18.8 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4753080  bytes 6866629048 (6.8 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

```
    runtime_env = {"env_vars": {"NCCL_SOCKET_IFNAME": "ens3"}}
    ray.init(runtime_env=runtime_env)
```

NOTE: Different hardware might require different values. g3 uses ens3 while g4dn uses ens5.

### Versions / Dependencies

master

### Reproduction script

https://sourcegraph.com/github.com/ray-project/ray/-/blob/release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py

Running pytorch_training_e2e.py with works across multiple hosts

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by amogkam**
Signed-off-by: Amog Kamsetty <amogkamsetty@yahoo.com>

Automatically set `NCCL_SOCKET_IFNAME` to prefer ethernet. Also adds a FAQ section in the docs on how to diagnose this issue.

Closes https://github.com/ray-project/ray/issues/26663

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/faq.rst [hunk 1]
```diff
@@ -26,3 +26,66 @@ you can initialize the ``Trainer`` with ``resources_per_worker`` specified in ``
...
+
+        veth526c8fe: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
+                inet6 fe80::44c:7bff:fe80:f02b  prefixlen 64  scopeid 0x20<link>
+                ether 06:4c:7b:80:f0:2b  txqueuelen 0  (Ethernet)
+                RX packets 24041  bytes 94697425 (94.6 MB)
+                RX errors 0  dropped 0  overruns 0  frame 0
+                TX packets 24062  bytes 2217752 (2.2 MB)
+                TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
+2. Choose the network interface that corresponds to the private IP address of your node. In most cases, this will be either
+   `ens3` or `ens5`.
```
Comment (matthewdeng) [inline doc/source/train/faq.rst:79]: Is there any NCCL documentation that corroborates this?

File: python/ray/train/constants.py [hunk 2]
```diff
@@ -79,3 +79,9 @@
 # By default these will be filtered out from ``session.report()``.
 # See ``TrainingCallback._preprocess_results`` for more details.
 ALL_RESERVED_KEYS = {PYTORCH_PROFILER_KEY}
+
+# Default NCCL_SOCKET_IFNAME.
+# Use ethernet when possible.
+# NCCL_SOCKET_IFNAME does a prefix match so "ens3" or "ens5" will match with
+# "en".
+DEFAULT_NCCL_SOCKET_IFNAME = "en,eth,bond"
```
Comment (matthewdeng) [inline python/ray/train/constants.py:87]: Double checking my understanding, does order matter here? Or does it matter what the order of `ifconfig` outputs?

**Review by matthewdeng — COMMENTED** [review]
Nice, thanks for doing this!!

File: doc/source/train/faq.rst (hunk 1 above)
Comment (amogkam) [reply doc/source/train/faq.rst:79]: no...nothing I could find. This is just from our experiences. Unfortunately the documentation around this is pretty sparse :(

File: python/ray/train/constants.py (hunk 2 above)
Comment (krfricke) [reply python/ray/train/constants.py:87]: The first is prefered (first anything starting in en, then eth, then bond)

**Review by krfricke — COMMENTED** [review]
LGTM, do we have passing release tests?

**Review by Yard1 — APPROVED** [review]
One question, should we log or warn when we are doing this?

**Comment by jiaodong** [comment]
Looks great ! Even updated tests are release tests with pytorch, can you also kick them off on PR to ensure everything works as expected ?

**Comment by amogkam** [comment]
@Yard1 added a debug statement

**Comment by amogkam** [comment]
Passing release test here: https://buildkite.com/ray-project/release-tests-pr/builds/15918#01835d56-ea64-4d75-8bac-56fc4d1c0198

**Review by jiaodong — APPROVED** [review]

**Comment by amogkam** [comment]
Failing tests are unrelated and are flaky. Going to merge.

**Comment by amogkam** [comment]
Torch comparison benchmark release test passing here: https://buildkite.com/ray-project/release-tests-pr/builds/15987#018361b3-f423-4f53-bfbf-ed73cf7a7f8d

**Pull request merged**